server/src/main/java/com/exacaster/lighter/backend/local/LocalApp.java

```diff
@@ -0,0 +1,95 @@
+package com.exacaster.lighter.backend.local;
+
+import static org.slf4j.LoggerFactory.getLogger;
+
+import com.exacaster.lighter.application.Application;
+import com.exacaster.lighter.application.ApplicationState;
+import com.exacaster.lighter.backend.SparkListener;
+import com.exacaster.lighter.backend.local.logger.LogCollectingHandler;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.function.Consumer;
+import java.util.logging.Logger;
+import org.apache.spark.launcher.SparkAppHandle;
+import org.apache.spark.launcher.SparkAppHandle.State;
+
+public class LocalApp implements SparkListener {
+    private static final org.slf4j.Logger LOG = getLogger(LocalApp.class);
+
+    private final LogCollectingHandler logHandle;
+    private final String loggerName;
+    private final Consumer<Throwable> errorHandler;
+    private State latestState;
```

Comment:
Maybe just use `handle.getState()` instead of this var? Seems like an additional reference to the state.